pkg/redshift/api/api.go

```diff
@@ -51,20 +65,120 @@ func (c *API) apiInput() apiInput {
 	return res
 }
 
-func (c *API) Execute(ctx aws.Context, query string) (*redshiftdataapiservice.ExecuteStatementOutput, error) {
+func (c *API) Execute(ctx context.Context, input *api.ExecuteQueryInput) (*api.ExecuteQueryOutput, error) {
 	commonInput := c.apiInput()
-	input := &redshiftdataapiservice.ExecuteStatementInput{
+	redshiftInput := &redshiftdataapiservice.ExecuteStatementInput{
 		ClusterIdentifier: commonInput.ClusterIdentifier,
 		Database:          commonInput.Database,
 		DbUser:            commonInput.DbUser,
 		SecretArn:         commonInput.SecretARN,
-		Sql:               aws.String(query),
+		Sql:               aws.String(input.Query),
+	}
+
+	output, err := c.Client.ExecuteStatementWithContext(ctx, redshiftInput)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", api.ExecuteError, err)
 	}
 
-	return c.Client.ExecuteStatementWithContext(ctx, input)
+	return &api.ExecuteQueryOutput{ID: *output.Id}, nil
+}
+
+func (c *API) Status(ctx aws.Context, output *api.ExecuteQueryOutput) (*api.ExecuteQueryStatus, error) {
+	statusResp, err := c.Client.DescribeStatementWithContext(ctx, &redshiftdataapiservice.DescribeStatementInput{
+		Id: aws.String(output.ID),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", api.StatusError, err)
+	}
+
+	var finished bool
+	state := *statusResp.Status
+	switch state {
+	case redshiftdataapiservice.StatusStringFailed,
+		redshiftdataapiservice.StatusStringAborted:
+		finished = true
+		err = errors.New(*statusResp.Error)
+	case redshiftdataapiservice.StatusStringFinished:
+		finished = true
+	default:
+		finished = false
+	}
+
+	return &api.ExecuteQueryStatus{
+		ID:       output.ID,
+		State:    state,
+		Finished: finished,
+	}, err
+}
+
+func (c *API) Stop(output *api.ExecuteQueryOutput) error {
+	_, err := c.Client.CancelStatement(&redshiftdataapiservice.CancelStatementInput{
+		Id: &output.ID,
+	})
+	if err != nil {
+		return fmt.Errorf("%w: %v", err, api.StopError)
+	}
+	return nil
+}
+
+func (c *API) Regions(aws.Context) ([]string, error) {
```

Comment:
I'm guessing there's no api call to aws to get this yeah? Listing it out manually helps to prevent/fix the same type of issue we were discussing yesterday for xray and ec2 permissions yeah?

Reply:
Recently I found an API that can give you the regions for a service: https://github.com/aws/aws-sdk-go/blob/main/aws/endpoints/endpoints.go#L360 but yeah, the hardcoded list here is a shortcut that has the benefit that does not need permissions.